packages/flutter/test/widgets/web_image_test.dart

```diff
@@ -0,0 +1,1308 @@
+    late bool isSync;
+    stream.addListener(WebImageStreamListener((WebImageInfo image, bool sync) {
+      isSync = sync;
+    }, (bool sync) {}));
+    expect(isSync, isTrue);
+  });
+
+  testWidgets('Precache removes original listener immediately after future completes, does not crash on successive calls #25143',
+  experimentalLeakTesting: LeakTesting.settings.withIgnoredAll(), // The test leaks by design, see [_TestImageStreamCompleter].
+  (WidgetTester tester) async {
+    final _TestWebImageStreamCompleter imageStreamCompleter = _TestWebImageStreamCompleter();
+    final _TestWebImageProvider provider = _TestWebImageProvider('example.com/pic.jpg', streamCompleter: imageStreamCompleter);
+
+    await tester.pumpWidget(
+      Builder(
+        builder: (BuildContext context) {
+          precacheWebImage(provider, context);
+          return Container();
+        },
+      ),
+    );
+
+    // Two listeners - one is the listener added by precacheImage, the other by the ImageCache.
+    final List<WebImageStreamListener> listeners = imageStreamCompleter.listeners.toList();
+    expect(listeners.length, 2);
+
+    // Make sure the first listener can be called re-entrantly
+    final WebImageInfo imageInfo = WebImageInfo(image10x10);
+
+    listeners[1].onImage(imageInfo.clone(), false);
+    listeners[1].onImage(imageInfo.clone(), false);
+
+    // Make sure the second listener can be called re-entrantly.
+    listeners[0].onImage(imageInfo.clone(), false);
+    listeners[0].onImage(imageInfo.clone(), false);
+
+    imageStreamCompleter.dispose();
+    imageCache.clear();
+  });
+
+  testWidgets('Precache completes with onError on error', (WidgetTester tester) async {
+    dynamic capturedException;
+    StackTrace? capturedStackTrace;
+    void errorListener(dynamic exception, StackTrace? stackTrace) {
+      capturedException = exception;
+      capturedStackTrace = stackTrace;
+    }
+
+    final Exception testException = Exception('cannot resolve host');
+    final StackTrace testStack = StackTrace.current;
+    final _TestWebImageProvider imageProvider = _TestWebImageProvider('example.com/pic.jpg');
+    late Future<void> precache;
+    await tester.pumpWidget(
+      Builder(
+        builder: (BuildContext context) {
+          precache = precacheWebImage(imageProvider, context, onError: errorListener);
+          return Container();
+        },
+      ),
+    );
+    imageProvider.fail(testException, testStack);
+    await precache;
+
+    // The image stream error handler should have the original exception.
+    expect(capturedException, testException);
+    expect(capturedStackTrace, testStack);
+    // If there is an error listener, there should be no FlutterError reported.
+    expect(tester.takeException(), isNull);
+  });
+
+  testWidgets('TickerMode controls stream registration', (WidgetTester tester) async {
+    final _TestWebImageStreamCompleter imageStreamCompleter = _TestWebImageStreamCompleter();
```

Comment:
Don't forget to dispose the `_TestWebImageStreamCompleter` instances in this test file, this one isn't cleaned up.

Reply:
Done.